[AMDGPU] Fold copies of constant physical registers into their uses

rampitec opened the pull request:
Co-authored-by: Jay Foad <Jay.Foad@amd.com>

rampitec committed:
[AMDGPU] Fold copies of constant physical registers into their uses

Co-authored-by: Jay Foad <Jay.Foad@amd.com>

rampitec commented:
* **#154410** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/154410?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/154410?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

rampitec committed:
Merge branch 'main' into users/rampitec/08-19-_amdgpu_fold_copies_of_constant_physical_registers_into_their_uses

jayfoad commented on llvm/test/CodeGen/AMDGPU/scale-offset-flat.ll:
```diff
@@ -412,11 +411,11 @@ define amdgpu_ps <2 x float> @flat_atomicrmw_b64_rtn_idxprom(ptr align 8 inreg %
+; GISEL-NEXT:    v_mov_b32_e32 v0, src_flat_scratch_base_lo
+; GISEL-NEXT:    v_cmp_ne_u64_e32 vcc_lo, 0, v[4:5]
+; GISEL-NEXT:    s_delay_alu instid0(VALU_DEP_2) | instskip(NEXT) | instid1(VALU_DEP_1)
+; GISEL-NEXT:    v_sub_nc_u32_e32 v0, v4, v0
```
Any idea why v0 didn't get folded into this sub?

rampitec replied:
I think I have some idea. Scalar is not considered acceptable here:
```
llvm-mc -arch=amdgcn -mcpu=gfx1250 <<< 'v_sub_nc_u32_e32 v0, v4, s0'
<stdin>:1:26: error: invalid operand for instruction
v_sub_nc_u32_e32 v0, v4, s0
                         ^
```

rampitec replied:
I.e., only in VOP3 form, but it is already VOP2 at the time of folding.

jayfoad replied:
> but it is already VOP2 at the time of folding

Oh, is this in the second run of folding, after shrinking?

Isn't there some logic to commute sub to subrev so that it can be folded?

Anyway this issue does not need to block this patch.

rampitec replied:
And yes, subrev shall be available. Which is a good potential optimization which we do not do there:
```
llvm-mc -arch=amdgcn -mcpu=gfx1250 -show-encoding <<< 'v_subrev_nc_u32 v0, src_flat_scratch_base_lo, v4'
	v_subrev_nc_u32_e32 v0, src_flat_scratch_base_lo, v4 ; encoding: [0xe6,0x08,0x00,0x4e]
```

jayfoad reviewed (commented):
LGTM but I wrote it, so it probably needs reviewing by someone else.

rampitec commented:
> LGTM but I wrote it, so it probably needs reviewing by someone else.

It was already reviewed twice, landed and reverted. What has changed: 1) I have fixed bug with aperture registers and got confirmation libc test now passing 2) I have updated the patch to factor physregs in one more place, where Matt has recently added some code.

Added Shilei.

jayfoad commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -727,8 +728,12 @@ bool SIFoldOperandsImpl::updateOperand(FoldCandidate &Fold) const {
+    Old.substPhysReg(New->getReg(), *TRI);
+  } else {
+    Old.substVirtReg(New->getReg(), New->getSubReg(), *TRI);
+    Old.setIsUndef(New->isUndef());
```
OK. But if it's physical then `New` should not be marked undef either, so I don't think it is actually _necessary_ to move this into the else.

rampitec replied:
Not even speaking that `setIsUndef` was placed as it was placed not deliberately, but as a result of a merge.

jayfoad replied:
I am not objecting to this change!
